src/main/java/org/prebid/server/bidder/openx/OpenxBidder.java

```diff
@@ -0,0 +1,231 @@
+                .map(BidderError::create)
+                .collect(Collectors.toList());
+    }
+
+    private List<HttpRequest<BidRequest>> createRequests(List<BidRequestWithError> bidRequestWithErrorsAll) {
+        return bidRequestWithErrorsAll.stream()
+                .map(BidRequestWithError::getBidRequest)
+                .filter(Objects::nonNull)
+                .map(singleBidRequest -> HttpRequest.of(HttpMethod.POST, endpointUrl,
+                        Json.encode(singleBidRequest), BidderUtil.headers(), singleBidRequest))
+                .collect(Collectors.toList());
+    }
+
+    private BidRequestWithError createSingleRequest(List<Imp> imps, BidRequest bidRequest) {
+
+        if (CollectionUtils.isEmpty(imps)) {
+            return BidRequestWithError.of(null, Collections.emptyList());
+        }
+
+        final List<String> errors = new ArrayList<>();
+        List<Imp> processedImps = null;
+
+        try {
+            processedImps = imps.stream().map(this::makeImp).collect(Collectors.toList());
+        } catch (PreBidException e) {
+            errors.add(e.getMessage());
+        }
+
+        final BidRequest ongoingRequest = CollectionUtils.isNotEmpty(processedImps)
+                ? bidRequest.toBuilder().imp(processedImps).ext(makeReqExt(imps)).build()
+                : null;
+
+        return BidRequestWithError.of(ongoingRequest, errors);
+    }
+
+    private Imp makeImp(Imp imp) {
+        final ExtImpOpenx openxImpExt = parseOpenxExt(imp);
+        return imp.toBuilder()
+                .tagid(openxImpExt.getUnit())
+                .bidfloor(openxImpExt.getCustomFloor())
+                .ext(makeImpExt(openxImpExt.getCustomParams()))
+                .build();
+    }
+
+    private ObjectNode makeReqExt(List<Imp> imps) {
+        return Json.mapper.valueToTree(OpenxRequestExt.of(parseOpenxExt(imps.get(0)).getDelDomain(), OPENX_CONFIG));
+    }
+
+    private ExtImpOpenx parseOpenxExt(Imp imp) {
+        final ObjectNode impExt = imp.getExt();
+        final ExtPrebid<?, ExtImpOpenx> impExtPrebid;
+        final ExtImpOpenx impExtOpenx;
+        if (impExt == null) {
+            throw new PreBidException("openx parameters section is missing");
+        }
+
+        try {
+            impExtPrebid = Json.mapper.convertValue(impExt, OPENX_EXT_TYPE_REFERENCE);
+            impExtOpenx = impExtPrebid != null ? impExtPrebid.getBidder() : null;
+            if (impExtOpenx == null) {
+                throw new PreBidException("openx parameters section is missing");
+            }
+        } catch (IllegalArgumentException e) {
+            logger.warn("Error occurred parsing openx parameters", e);
+            throw new PreBidException(e.getMessage(), e);
+        }
+        return impExtOpenx;
+    }
+
+    private ObjectNode makeImpExt(Map<String, String> customParams) {
+        return customParams != null
+                ? Json.mapper.valueToTree(OpenxImpExt.of(Collections.unmodifiableMap(customParams)))
```

Comment:
I do not understand why unmodifiableMap() was used here? Maybe, if it is important, we can left comment here, otherwise remove this call

Reply:
It's a stale expression. Thanks for attentiveness.